core-bundle/src/Routing/ResponseContext/ResponseContext.php

```diff
@@ -12,31 +12,46 @@
 
 namespace Contao\CoreBundle\Routing\ResponseContext;
 
-use Symfony\Component\HttpFoundation\Response;
-use Symfony\Component\HttpFoundation\ResponseHeaderBag;
-
-class ResponseContext implements ResponseContextInterface
+final class ResponseContext
 {
+    public const REQUEST_ATTRIBUTE_NAME = '_contao_response_context';
+
+    /**
+     * @var array
+     */
+    private $services = [];
+
     /**
      * @var PartialResponseHeaderBag|null
      */
     private $headerBag;
 
-    public function getHeaderBag(): ResponseHeaderBag
+    public function add(object $service): self
     {
-        if (null === $this->headerBag) {
-            $this->headerBag = new PartialResponseHeaderBag();
-        }
+        $this->services[\get_class($service)] = $service;
 
-        return $this->headerBag;
+        return $this;
+    }
+
+    public function has(string $service): bool
+    {
+        return null !== $this->get($service);
     }
 
-    public function finalize(Response $response): ResponseContextInterface
+    /**
+     * @return mixed|null
```

Comment:
This wants some generics magic :smile: (also see next comment why I'd also remove `null`):

```suggestion
     * @template T
     * @psalm-param class-string<T> $service
     * @psalm-return T
     *
     * @return object
```